### ISSUE TITLE
Machine parsable migration attempts

### ISSUE THREAD

**Issue by nthykier** [Debian/britney2#3]
Raphaël Hertzog suggested that it would be nice to have a machine parsable file of the migration attempts (instead of or in addition to the current output.log).  Requested features include:

- [ ] Separate packages that are directly broken from packages that are transitively broken.  (Caveat: not all break happens at "first" level).
- [ ] In the list of old libraries, include a list of (versioned) reverse dependencies that prevents their removal.

**Comment by cyphermox** [comment]
I've been writing a script to parse the additional excuses.yaml generated in Ubuntu's "fork" of britney, and getting towards wanting to parse output too; it would be indeed nice to have a machine-parseable update_output. I don't know about directly broken vs. transitively broken, but as a first step output could be in (say, YAML, which I've mocked here: https://paste.ubuntu.com/p/c53gXp2hRg/) a machine-readable format.

Rather than making further changes to Ubuntu only to support this, and since I tried to keep the example above as close as possible to the current output format so it's not completely foreign to its current users, how do you feel about this change?

This would just be changing the string passed to output.write() for now, to keep the outputting as quick as possible.

**Comment by nthykier** [comment]
Mathieu Trudel-Lapierre:
> I've been writing a script to parse the additional excuses.yaml generated in Ubuntu's "fork" of britney, and getting towards wanting to parse output too; it would be indeed nice to have a machine-parseable update_output. I don't know about directly broken vs. transitively broken, but as a first step output could be in (say, YAML, which I've mocked here: https://paste.ubuntu.com/p/c53gXp2hRg/) a machine-readable format.
> 
> Rather than making further changes to Ubuntu only to support this, and since I tried to keep the example above as close as possible to the current output format so it's not completely foreign to its current users, how do you feel about this change?
> 
> This would just be changing the string passed to output.write() for now, to keep the outputting as quick as possible.
> 

Thanks for considering to improve britney upstream. :)


The issue is that "update_output" is basically a log file and meant
primarily to be "human readable" (even if it could do with some
improvements in that area - at least for newcomers to britney).
Furthermore, given it is a log output, most commiters will probably not
think twice with changing message layouts or content.

Therefore, I think we are better served with a separate output that is
from start to end a machine-readable output.  Can you say what do you
need the data for / what data do you want available in a
machine-readable output?

Thanks,
~Niels

**Comment by cyphermox** [comment]
On Tue, Mar 13, 2018 at 9:41 PM, Niels Thykier <notifications@github.com>
wrote:

> The issue is that "update_output" is basically a log file and meant
> primarily to be "human readable" (even if it could do with some
> improvements in that area - at least for newcomers to britney).
> Furthermore, given it is a log output, most commiters will probably not
> think twice with changing message layouts or content.
>
Therefore, I think we are better served with a separate output that is
> from start to end a machine-readable output.
>

I'm not attached to either way. When I ran this by Steve Langasek for
Ubuntu's britney, it was suggested that having a separate object to build
say, YAML would potentially negatively impact the performance of britney.
It's not quite the same issue, but my initial idea was to do the
implementation with as little change as possible.


> Can you say what do you need the data for / what data do you want
> available in a
> machine-readable output?


Right now, my script can provide a maintainer or someone interested in
helping in the maintenance of the devel release with what would be the next
work needed to have a package transition (in Ubuntu's case, from -proposed
to -release). It's much like listing excuses, with additional parsing to
provide with actionable tasks for more info: link to build log, whether
something is a NBS, architecture removed from debian/control, etc. All of
this can be done from the command-line with a machine-parseable output.

The next logical step is to apply similar logic to unblocking transitions,
although that uses the output of britney instead of the excuses: for each
case in "Trying easy", for instance, setting up a cache that includes both
the current archive and the new packages and attempting to install the new
packages, to check for what breaks instability of the affected packages.
Parsing the log in its current form isn't simple to humans, there's often
confusion in the meaning of the contents of update_output.txt, and it's not
much better to do programmatically in its current form. To do this, I'd
need the 'trying easy' lines along with the breakdown per architecture.
Anything else is not critical to my current project, but I thought it a
good idea to have the full output, in case there is more useful information
to extract at a later point.

I was asking this here seeing as the bug mentions machine-parseable output
in general, and I'm happy to do the work if we agree on the design and it
can benefit both Debian and Ubuntu.

Incidentally, is there a more suitable place to discuss this than github
issues?


[1] https://git.launchpad.net/~cyphermox/ubuntu-archive-assistant/
[2]
http://people.canonical.com/~ubuntu-archive/proposed-migration/update_excuses.yaml

Mathieu Trudel-Lapierre <mathieu.tl@gmail.com>
Freenode: cyphermox, Jabber: mathieu.tl@gmail.com
4096R/65B58DA1 818A D123 0992 275B 23C2  CF89 C67B B4D6 65B5 8DA1

**Comment by nthykier** [comment]
Mathieu Trudel-Lapierre:
> On Tue, Mar 13, 2018 at 9:41 PM, Niels Thykier <notifications@github.com>
> wrote:
> 
>> The issue is that "update_output" is basically a log file and meant
>> primarily to be "human readable" (even if it could do with some
>> improvements in that area - at least for newcomers to britney).
>> Furthermore, given it is a log output, most commiters will probably not
>> think twice with changing message layouts or content.
>>
>> Therefore, I think we are better served with a separate output that is
>> from start to end a machine-readable output.
>>
> 
> I'm not attached to either way. When I ran this by Steve Langasek for
> Ubuntu's britney, it was suggested that having a separate object to build
> say, YAML would potentially negatively impact the performance of britney.

You cannot avoid building up the YAML file in a large big object that
gets flushed in one go with most YAML writers I know.  However, by
separating the log file and the machine data, we can at least avoid also
having to cache the entire log file as well.

> It's not quite the same issue, but my initial idea was to do the
> implementation with as little change as possible.
> 

Ok.  We will have to find the proper trade-off between how much change
is needed vs. what data can we extract in the short term vs. what data
quality do we get (in the long term).

> 
>> Can you say what do you need the data for / what data do you want
>> available in a
>> machine-readable output?
> 
> 
> Right now, my script can provide a maintainer or someone interested in
> helping in the maintenance of the devel release with what would be the next
> work needed to have a package transition (in Ubuntu's case, from -proposed
> to -release). It's much like listing excuses, with additional parsing to
> provide with actionable tasks for more info: link to build log, whether
> something is a NBS, architecture removed from debian/control, etc. All of
> this can be done from the command-line with a machine-parseable output.
> 

Can you show some examples of its output and say a little more about how
it is used in Ubuntu?  It sounds like it might be useful to move some of
that functionality into the britney codebase at some point (maybe
britney itself, maybe in a separate script).

Possible related note: Britney has gotten some documentation now that
might be worth linking to from that tool.  The documentation is in the
doc folder the Britney source (for when you need to add something)

Debian hosts that documentation on https://release.debian.org/doc/britney/

> The next logical step is to apply similar logic to unblocking transitions,
> although that uses the output of britney instead of the excuses: for each
> case in "Trying easy", for instance, setting up a cache that includes both
> the current archive and the new packages and attempting to install the new
> packages, to check for what breaks instability of the affected packages.

Are you aware that transitions do not have a 1:1 with easy hints?  It is
a 1:0..N relation.  Notably, britney does "multi-item migrations" in the
main run sometimes (these work similar to easy hints but are not
classified as such).

Often, britney will perform the success migration of the transition in
the main-run as a "non-easy" multi-item migration.

> Parsing the log in its current form isn't simple to humans, there's often
> confusion in the meaning of the contents of update_output.txt, and it's not
> much better to do programmatically in its current form.

It can certainly be done better; my primary issue with it (from a human
perspective) is the learning curve.  For machine-readable, then it is
not really optimized for it (and could use a standard format)

> To do this, I'd
> need the 'trying easy' lines along with the breakdown per architecture.

What about (non-easy) multi-item migrations?

> Anything else is not critical to my current project, but I thought it a
> good idea to have the full output, in case there is more useful information
> to extract at a later point.
> 

Ok, I am mostly concerned with the form ("quality") of the output and
our ability to extend the format without committing ourselves too much
to "uninteresting" information.

> I was asking this here seeing as the bug mentions machine-parseable output
> in general, and I'm happy to do the work if we agree on the design and it
> can benefit both Debian and Ubuntu.
> 
> Incidentally, is there a more suitable place to discuss this than github
> issues?
> 

I think this place is fine.  However, it is not the most official
hosting for Britney.  The (most) official place would be filing a bug
against release.debian.org and move the discussion there.  But as I
opened with, for me either location is fine.

Note that there are a larger number of people reading the Debian bugs
(incl. some that mostly works with other areas of Debian release work).
If you want to move the discussion, then we should keep some context for
them (as to avoid confusing them).
